src/game/carrier.c

```diff
@@ -0,0 +1,213 @@
+
+static const LEGACY_DROP_TEST m_LegacyMap[] = {
+    { O_PIERRE, O_SCION_ITEM2 },
+    { O_COWBOY, O_MAGNUM_ITEM },
+    { O_SKATEKID, O_UZI_ITEM },
+    { O_BALDY, O_SHOTGUN_ITEM },
+};
+
+void Carrier_InitialiseLevel(int32_t level_num)
+{
+    int32_t total_item_count = g_LevelItemCount;
+    GAMEFLOW_LEVEL level = g_GameFlow.levels[level_num];
+    for (int i = 0; i < level.item_drops.count; i++) {
+        GAMEFLOW_DROP_ITEM_DATA *data = &level.item_drops.data[i];
+
+        ITEM_INFO *item = Carrier_GetCarrier(data->enemy_num);
+        if (!item) {
+            LOG_WARNING("%d does not refer to a loaded item", data->enemy_num);
+            continue;
+        }
+
+        if (total_item_count + data->count > MAX_ITEMS) {
+            LOG_WARNING("Too many items being loaded");
+            return;
+        }
+
+        if (item->carried_item) {
+            LOG_WARNING("Item %d is already carrying", data->enemy_num);
+            continue;
+        }
+
+        if (!Carrier_IsObjectType(item->object_number, m_CarrierObjects)) {
+            LOG_WARNING(
+                "Item %d of type %d cannot carry items", data->enemy_num,
+                item->object_number);
+            continue;
+        }
+
+        bool valid_objects = true;
+        for (int i = 0; i < data->count; i++) {
+            if (!Carrier_IsObjectType(data->object_ids[i], m_DropObjects)) {
+                LOG_WARNING(
+                    "Items of type %d cannot be carried", data->object_ids[i]);
+                valid_objects = false;
+                break;
+            }
+        }
+
+        if (!valid_objects) {
+            continue;
+        }
+
+        item->carried_item =
+            GameBuf_Alloc(sizeof(CARRIED_ITEM) * data->count, GBUF_ITEMS);
+        CARRIED_ITEM *drop = item->carried_item;
+        for (int i = 0; i < data->count; i++) {
+            drop->object_id = data->object_ids[i];
+            drop->spawn_number = NO_ITEM;
+            drop->status = IS_NOT_ACTIVE;
+            total_item_count++;
+
+            if (i < data->count - 1) {
+                drop->next_item = drop + 1;
+                drop++;
+            } else {
+                drop->next_item = NULL;
+            }
+        }
+    }
+}
+
+static ITEM_INFO *Carrier_GetCarrier(int16_t item_num)
+{
+    if (item_num < 0 || item_num >= g_LevelItemCount) {
+        return NULL;
+    }
+
+    ITEM_INFO *item = &g_Items[item_num];
+    if (Carrier_IsObjectType(item->object_number, m_PlaceholderObjects)) {
+        int16_t child_item_num = *(int16_t *)item->data;
+        item = &g_Items[child_item_num];
+    }
+
+    if (!g_Objects[item->object_number].loaded) {
+        return NULL;
+    }
+
+    return item;
+}
+
+static bool Carrier_IsObjectType(int16_t object_id, const int16_t *test_arr)
+{
+    for (int i = 0; test_arr[i] != NO_ITEM; i++) {
+        if (test_arr[i] == object_id) {
+            return true;
+        }
+    }
+    return false;
+}
```

Comment:
Can the `int16_t` throughout this file become `GAME_OBJECT_ID`?

Reply:
Yes, that would be much neater I think.